File: app/src/main/java/com/timotiusoktorio/pencake/ui/profile/ProfileViewModel.java
package com.timotiusoktorio.pencake.ui.profile;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.timotiusoktorio.pencake.data.model.User;
import com.timotiusoktorio.pencake.data.source.DataManager;

class ProfileViewModel extends ViewModel {

    final MutableLiveData<User> userLiveData = new MutableLiveData<>();
    final MutableLiveData<Boolean> updatePassSuccessFlagLiveData = new MutableLiveData<>();
    final MutableLiveData<String> updatePassErrorMsgLiveData = new MutableLiveData<>();
    final MutableLiveData<Boolean> signOutSuccessFlagLiveData = new MutableLiveData<>();

    private final DataManager dataManager;

    ProfileViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
        loadUser();
    }

    private void loadUser() {
        dataManager.fetchUser(new DataManager.SingleCallback<User>() {
            @Override
            public void onSuccess(User data) {
                userLiveData.setValue(data);
            }

            @Override
            public void onError(String errorMsg) {

            }
        });
    }

    void updateUser(String name, String phone) {
        User user = userLiveData.getValue();
        if (user != null) {
            user.setDisplayName(name);
            user.setPhone(phone);
            dataManager.updateUser(user);
        }
    }

    void updatePassword(String oldPassword, final String newPassword) {
        dataManager.updatePassword(oldPassword, newPassword, new DataManager.SingleCallback<Boolean>() {
            @Override
            public void onSuccess(Boolean data) {
                updatePassSuccessFlagLiveData.setValue(data);
            }

            @Override
            public void onError(String errorMsg) {
                updatePassErrorMsgLiveData.setValue(errorMsg);
            }
        });
    }

    void signOut() {
        FirebaseAuth.getInstance().signOut();
        dataManager.removeFavoriteList();
        signOutSuccessFlagLiveData.setValue(true);
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/profile/ProfileViewModel.kt
package com.timotiusoktorio.pencake.ui.profile

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.timotiusoktorio.pencake.data.model.User
import com.timotiusoktorio.pencake.data.source.DataManager

class ProfileViewModel(private val dataManager: DataManager) : ViewModel() {

    val userLiveData = MutableLiveData<User>()
    val updatePassSuccessFlagLiveData = MutableLiveData<Boolean>()
    val updatePassErrorMsgLiveData = MutableLiveData<String>()
    val signOutSuccessFlagLiveData = MutableLiveData<Boolean>()

    init {
        loadUser()
    }

    private fun loadUser() {
        dataManager.fetchUser(object : DataManager.SingleCallback<User> {
            override fun onSuccess(data: User) {
                userLiveData.value = data
            }

            override fun onError(errorMsg: String) {
                // Not implemented
            }
        })
    }

    fun updateUser(name: String, phone: String) {
        userLiveData.value?.let {
            it.displayName = name
            it.phone = phone
            dataManager.updateUser(it)
        }
    }

    fun updatePassword(oldPassword: String, newPassword: String) {
        dataManager.updatePassword(oldPassword, newPassword, object : DataManager.SingleCallback<Boolean> {
            override fun onSuccess(data: Boolean) {
                updatePassSuccessFlagLiveData.value = data
            }

            override fun onError(errorMsg: String) {
                updatePassErrorMsgLiveData.value = errorMsg
            }
        })
    }

    fun signOut() {
        FirebaseAuth.getInstance().signOut()
        signOutSuccessFlagLiveData.value = true
    }
}